### PR TITLE
Take more care over HostConfig on start request in proxy

### LINKED ISSUE #1531: weaveproxy can overwrite port bindings from docker-compose
To recreate using `docker-compose`, create a file `docker-compose.yml`:

    foo:
      image: gliderlabs/alpine
      command: sleep 100
      ports:
       - "1234:1234"

Start this container via `weaveproxy`:

    DOCKER_HOST=unix:///var/run/weave/weave.sock docker-compose up -d

Check for the published port:

    $ docker inspect -f '{{.HostConfig.PortBindings}}' test_foo_1
    map[]

Cause is that `docker-compose` is sending in an empty JSON request on `start`; we are setting our HostConfig fields on that request, and hence overwriting the `PortBindings`.

This does not happen if you try via `docker create` and `docker start`, since `docker start` does not send any content at all in the request and `startContainerInterceptor.InterceptRequest()` detects that and bypasses setting the HostConfig.

### PR DESCRIPTION
**Pull request by bboreham**
"There are more things in heaven and earth, `HostConfig`, than are dreamt of in your philosophy"

Fixes #1531 and #1532

### CODE REVIEW

File: test/666_abuse_of_start_test.sh [hunk 1]
```diff
@@ -11,7 +11,12 @@ proxy_start_container $HOST1 --name=c1
 # Start c2 with a sneaky HostConfig
 proxy docker_on $HOST1 create --name=c2 $SMALL_IMAGE $CHECK_ETHWE_UP
 proxy docker_api_on $HOST1 POST /containers/c2/start '{"NetworkMode": "container:c1"}'
-assert "docker_on $HOST1 inspect -f '{{.State.Running}} {{.State.ExitCode}} {{.HostConfig.Dns}}' c2" "false 0 [$docker_bridge_ip]"
+assert "docker_on $HOST1 inspect -f '{{.HostConfig.NetworkMode}} {{.State.Running}} {{.State.ExitCode}} {{.HostConfig.Dns}}' c2" "container:c1 false 0 [$docker_bridge_ip]"
+
+# Start c5 with a differently sneaky HostConfig
+proxy docker_on $HOST1 create --name=c5 $SMALL_IMAGE $CHECK_ETHWE_UP
+proxy docker_api_on $HOST1 POST /containers/c5/start '{"HostConfig": {"NetworkMode": "container:c1"}}'
+assert "docker_on $HOST1 inspect -f '{{.HostConfig.NetworkMode}} {{.State.Running}} {{.State.ExitCode}} {{.HostConfig.Dns}}' c5" "container:c1 false 0 [$docker_bridge_ip]"
```
Comment (rade) [inline test/666_abuse_of_start_test.sh]: The repetition of the `{{.State.Running}} {{.State.ExitCode}} {{.HostConfig.Dns}}` and `false 0 [$docker_bridge_ip]` is getting a bit tedious. I did bite my tongue when there were only two of these. Now there are three.

Comment (bboreham) [inline test/666_abuse_of_start_test.sh]: The third one is different. I have pushed a change to reduce the repetition.

**Comment by paulbellamy** [comment]
If the user sends *both* a top-level HostConfig, *and* a named sub-HostConfig, it looks like we only modify the inner one. Will that override the outer?

**Comment by rade** [comment]
needs rebasing - the new `attach` should presumably go into the `check_...` function.

**Comment by paulbellamy** [comment]
Maybe add a test for cpuset handling, and the case where there are both HostConfigs, but other than that, LGTM.

**Comment by bboreham** [comment]
I looked into `Cpuset`.  Docker only uses that value if you send _both_ `Cpuset` and `HostConfig`.  So, for any caller sending _just_ `Cpuset`, we change the behaviour: previously docker would just ignore the `Cpuset` and now it will use it, and additionally any `HostConfig` parameters supplied in the original create will be blanked out.

However, this undesirable behaviour will happen for any valid JSON that Docker was previously ignoring, and we cannot detect all such requests.  So I propose not to detect any of them.

File: test/666_abuse_of_start_test.sh [hunk 2]
```diff
@@ -4,24 +4,38 @@
 
 start_suite "Abuse of 'start' operation"
 
-weave_on $HOST1 launch
+weave_on $HOST1 launch --log-level=debug
```
Comment (rade) [inline test/666_abuse_of_start_test.sh]: I'm guessing that's not meant to be there.

File: test/666_abuse_of_start_test.sh [hunk 3]
```diff
@@ -4,24 +4,38 @@
...
 docker_bridge_ip=$(weave_on $HOST1 docker-bridge-ip)
 proxy_start_container $HOST1 --name=c1
 
+check_hostconfig() {
+    docker_on $HOST1 attach c2 >/dev/null || true # Wait for container to exit
```
Comment (rade) [inline test/666_abuse_of_start_test.sh]: on master this now suppresses errors with `>/dev/null 2>&1`. Should do here too.

File: proxy/start_container_interceptor.go [hunk 4]
```diff
@@ -18,25 +18,33 @@ func (i *startContainerInterceptor) InterceptRequest(r *http.Request) error {
...
-	if err := marshalRequestBody(r, hostConfig); err != nil {
-		return err
+		// HostConfig can be sent either unnamed at top level or as a struct named HostConfig
+		hostConfig := params
+		if subParam, found := params["HostConfig"]; found {
+			if typecast, ok := subParam.(map[string]interface{}); ok {
+				hostConfig = typecast
+			}
+		}
+		if len(hostConfig) > 0 {
```
Comment (paulbellamy) [inline proxy/start_container_interceptor.go]: What about `{"HostConfig":{}}`, looks like that won't trigger this?

Comment (bboreham) [inline proxy/start_container_interceptor.go]: I think that will leave both pointers as nil so do nothing.  Unless you've tested it?

**Comment by rade** [comment]
@bboreham needs rebase, apparently.

**Comment by bboreham** [comment]
Rebased and squashed into fewer commits

File: proxy/start_container_interceptor.go (hunk 4 above)
Comment (paulbellamy) [inline proxy/start_container_interceptor.go]: Have pushed a failing test for it.

**Pull request merged**